File: _lcd4.py
from ._stuff import *

import collections


def lcd4(
    *, prefix, data_pindesc4567, e_pindesc, rs_pindesc, header=True
):
    assert len(data_pindesc4567)==4
    assert data_pindesc4567[0].port == data_pindesc4567[1].port == data_pindesc4567[2].port == data_pindesc4567[3].port
    dico = collections.OrderedDict()
    dico["LCDPREFIX"] = prefix

    filename = LIBDIR + "/lcd4." + ("h" if header else "c")
    with open(filename) as f:
        content = f.read()

    for k, v in dico.items():
        content = content.replace(k, v)

    filtered = content.splitlines()

    while filtered and not filtered[0]:
        del filtered[0]
    while filtered and not filtered[-1]:
        del filtered[-1]

    content = "\n".join(filtered)

    constants = collections.OrderedDict(
        {
            "LCD_DATA_PORT": data_pindesc4567[0].port,
            "LCD_DATA_PIN4": data_pindesc4567[0].pin_bm,
            "LCD_DATA_PIN5": data_pindesc4567[1].pin_bm,
            "LCD_DATA_PIN6": data_pindesc4567[2].pin_bm,
            "LCD_DATA_PIN7": data_pindesc4567[3].pin_bm,
            "LCD_E_PORT": e_pindesc.port,
            "LCD_E_PIN": e_pindesc.pin_bm,
            "LCD_RS_PORT": rs_pindesc.port,
            "LCD_RS_PIN": rs_pindesc.pin_bm,
            "LCD_CMD_LINEONE": "0x00",
            "LCD_CMD_LINETWO": "0x40",
            "LCD_CMD_CLEAR": "0b00000001",
            "LCD_CMD_HOME": "0b00000010",
            "LCD_CMD_ENTRYMODE": "0b00000110",
            "LCD_CMD_DISPLAYOFF": "0b00001000",
            "LCD_CMD_DISPLAYON": "0b00001100",
            "LCD_CMD_FUNCTIONRESET": "0b00110000",
            "LCD_CMD_FUNCTIONSET4BIT": "0b00101000",
            "LCD_CMD_SETCURSOR": "0b10000000",
        }
    )

    defined, content = generate_define("LCD", prefix, constants, content)

    if header:
        return protect_header(f"LCD4_{prefix}", defined + "\n" + content)
    return content

File: libs/lcd4.c
void LCDPREFIX_init()
{
    LCD_DATA_PORT.DIRSET = LCD_DATA_PIN7 | LCD_DATA_PIN6 | LCD_DATA_PIN5 | LCD_DATA_PIN4;
    LCD_E_PORT.DIRSET = LCD_E_PIN;
    LCD_RS_PORT.DIRSET = LCD_RS_PIN;

    _delay_ms(100);

    LCD_RS_PORT.OUTCLR = LCD_RS_PIN;
    LCD_E_PORT.OUTCLR = LCD_E_PIN;

    LCDPREFIX_write(LCD_CMD_FUNCTIONRESET);
    _delay_ms(10);

    LCDPREFIX_write(LCD_CMD_FUNCTIONRESET);
    _delay_us(200);

    LCDPREFIX_write(LCD_CMD_FUNCTIONRESET);
    _delay_us(200);
 
    LCDPREFIX_write(LCD_CMD_FUNCTIONSET4BIT);
    _delay_us(80);

    LCDPREFIX_write_instruction(LCD_CMD_FUNCTIONSET4BIT);
    _delay_us(80);

    LCDPREFIX_write_instruction(LCD_CMD_DISPLAYOFF);
    _delay_us(80);

    LCDPREFIX_write_instruction(LCD_CMD_CLEAR);
    _delay_ms(4);

    LCDPREFIX_write_instruction(LCD_CMD_ENTRYMODE);
    _delay_us(80);

    LCDPREFIX_write_instruction(LCD_CMD_DISPLAYON);
    _delay_us(80);
LCDCUSTOMCHARS}

void LCDPREFIX_write_zchar(char* what)
{
    LCDPREFIX_write_instruction(LCD_CMD_CLEAR);
    _delay_ms(4);
    
    LCDPREFIX_write_instruction(LCD_CMD_SETCURSOR | LCD_CMD_LINEONE);
    _delay_us(80);

    uint8_t i=0;
    for(i=0; i<16; i++)
    {
        if(what[i]==0)
            break;
        if(what[i]=='\n')
            break;
        LCDPREFIX_write_char(what[i]);
        _delay_us(80);
    }
    
    if(what[i]!=0)
    {
        LCDPREFIX_write_instruction(LCD_CMD_SETCURSOR | LCD_CMD_LINETWO);
        _delay_us(80);
        i++;
        uint8_t istart = i;
        while(i-istart<16)
        {
            if(what[i]==0)
                break;
            if(what[i]=='\n')
                break;
            LCDPREFIX_write_char(what[i]);
            _delay_us(80);
            i++;
        }
    }
}


void LCDPREFIX_write_lines(char* line0, char* line1)
{
    LCDPREFIX_write_instruction(LCD_CMD_CLEAR);
    _delay_ms(4);

    LCDPREFIX_write_instruction(LCD_CMD_SETCURSOR | LCD_CMD_LINEONE);
    _delay_us(80);

    LCDPREFIX_write_a_line(line0);
    
    LCDPREFIX_write_instruction(LCD_CMD_SETCURSOR | LCD_CMD_LINETWO);
    _delay_us(80);
    
    LCDPREFIX_write_a_line(line1);
}


void LCDPREFIX_write_a_line(char* what)
{
    uint8_t i=0;
    for(i=0; i<16; i++)
    {
        if(what[i]==0)
            break;
        if(what[i]=='\n')
            break;
        LCDPREFIX_write_char(what[i]);
        _delay_us(80);
    }
}   


void LCDPREFIX_write_char(uint8_t what)
{
    LCD_RS_PORT.OUTSET = LCD_RS_PIN;
    LCD_E_PORT.OUTCLR = LCD_E_PIN;
    LCDPREFIX_write(what);
    LCDPREFIX_write(what << 4);
}

void LCDPREFIX_write_instruction(uint8_t what)
{
    LCD_RS_PORT.OUTCLR = LCD_RS_PIN;
    LCD_E_PORT.OUTCLR = LCD_E_PIN;
    LCDPREFIX_write(what);
    LCDPREFIX_write(what << 4);
}

void LCDPREFIX_write(uint8_t what)
{
    LCD_DATA_PORT.OUTCLR = LCD_DATA_PIN7 | LCD_DATA_PIN6 | LCD_DATA_PIN5 | LCD_DATA_PIN4;
    uint8_t outset = 0x00;
    if(what & (1<<7))
        outset |= LCD_DATA_PIN7;
    if(what & (1<<6))
        outset |= LCD_DATA_PIN6;
    if(what & (1<<5))
        outset |= LCD_DATA_PIN5;
    if(what & (1<<4))
        outset |= LCD_DATA_PIN4;
    LCD_DATA_PORT.OUTSET = outset;

    LCD_E_PORT.OUTSET = LCD_E_PIN;
    _delay_us(1);
    LCD_E_PORT.OUTCLR = LCD_E_PIN;
    _delay_us(1);
}
